From 696a5e6483baa3a9e4eb83e5db5838d571c90ec9 Mon Sep 17 00:00:00 2001
From: Lee Duncan <lduncan@suse.com>
Date: Sun, 14 Feb 2021 14:59:07 -0800
Subject: [PATCH libaio] Fix test issue with gcc-11
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Test 3.t exercises io_submit() with invalid addresses,
and one test attempted to pass in "-1" as an invalid
pointer. But gcc-11 realizes you cannot take an offset
from an invalid pointer, making it impossible to
compile or run this test:

> sh# make CC=gcc-11 partcheck
> make[1]: Entering directory '/alt/public_software/libaio/src'
> make[1]: Nothing to be done for 'all'.
> make[1]: Leaving directory '/alt/public_software/libaio/src'
> make[1]: Entering directory '/alt/public_software/libaio/harness'
> gcc-11 -Wall -Werror -I../src -g -O2 -DTEST_NAME=\"cases/2.t\" -o cases/2.p main.c ../src/libaio.a -lpthread
> gcc-11 -Wall -Werror -I../src -g -O2 -DTEST_NAME=\"cases/3.t\" -o cases/3.p main.c ../src/libaio.a -lpthread
> In file included from main.c:24:
> cases/3.t: In function ‘test_main’:
> cases/3.t:18:19: error: ‘attempt_io_submit’ accessing 8 bytes in a region of size 0 [-Werror=stringop-overflow=]
>    18 |         status |= attempt_io_submit( io_ctx, 1, (void *)-1, -EFAULT);
>       |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> cases/3.t:18:19: note: referencing argument 3 of type ‘struct iocb **’
> In file included from cases/3.t:5,
>                  from main.c:24:
> cases/aio_setup.h:15:5: note: in a call to function ‘attempt_io_submit’
>    15 | int attempt_io_submit(io_context_t ctx, long nr, struct iocb *ios[], int expect)
>       |     ^~~~~~~~~~~~~~~~~
> cc1: all warnings being treated as errors
> make[1]: *** [Makefile:19: cases/3.p] Error 1
> make[1]: Leaving directory '/alt/public_software/libaio/harness'
> make: *** [Makefile:23: partcheck] Error 2

The fix is to tell gcc to relax it's stringop overflow
testing for this one test.

Changes since v1:
* First version just skipped the test in question

Signed-off-by: Lee Duncan <leeman.duncan@gmail.com>
Signed-off-by: Jeff Moyer <jmoyer@redhat.com>
---
 harness/Makefile | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/harness/Makefile b/harness/Makefile
index 5cc2b25..4f225d3 100644
--- a/harness/Makefile
+++ b/harness/Makefile
@@ -9,6 +9,11 @@ HARNESS_SRCS:=main.c
 CFLAGS+=-Wall -Werror -I../src -g -O2
 #-lpthread -lrt
 
+# gcc-11 does not like the test case in 3.t that
+# passes an invalid pointer (-1) to the kernel, so
+# tell the compiler we do not need a warning here
+cases/3.p:	CFLAGS+=-Wno-stringop-overflow
+
 # Change this on the build line to run tests against the installed libraries:
 # make LIBAIO=-laio partcheck
 LIBAIO?=../src/libaio.a
-- 
2.33.1

